File: sdks/typescript/src/llm-adapter.ts
/**
 * LLM adapter that wraps any provider to speak the durable protocol.
 * Zero dependencies — Node.js built-ins only.
 */

import { createMessageReader, writeMessage, ProtocolMessage } from "./protocol";
import { Message } from "./types";

type LlmHandler = (
  messages: Message[],
  tools?: unknown,
  model?: string
) => Promise<{ content?: string; text?: string; tool_calls?: unknown[] }>;

/**
 * Wrap an LLM provider to speak the durable NDJSON protocol on stdio.
 *
 * @example
 * ```ts
 * const adapter = new DurableLlmAdapter(async (messages, tools, model) => {
 *   const response = await openai.chat.completions.create({ messages, tools, model });
 *   return response.choices[0].message;
 * });
 * adapter.run();
 * ```
 */
export class DurableLlmAdapter {
  constructor(private handler: LlmHandler) {}

  async run(): Promise<void> {
    for await (const msg of createMessageReader()) {
      if (msg.type === "chat") {
        const messages = (msg.messages as Message[]) || [];
        const tools = msg.tools;
        const model = msg.model as string | undefined;
        await this.handleChat(messages, tools, model);
      } else if (msg.type === "heartbeat") {
        writeMessage({ type: "heartbeat_ack", timestamp: msg.timestamp });
      } else {
        writeMessage({
          type: "error",
          message: `unknown message type: ${msg.type}`,
          retryable: false,
        });
      }
    }
  }

  private async handleChat(
    messages: Message[],
    tools: unknown,
    model?: string
  ): Promise<void> {
    try {
      const result = await this.handler(messages, tools, model);

      if (result.tool_calls && result.tool_calls.length > 0) {
        writeMessage({ type: "tool_calls", calls: result.tool_calls });
      } else {
        const content = result.content || result.text || "";
        writeMessage({ type: "text", content });
      }
    } catch (err: unknown) {
      writeMessage({
        type: "error",
        message: err instanceof Error ? err.message : String(err),
        retryable: true,
      });
    }
  }
}

File: sdks/python/durable_sdk/llm_adapter.py
"""LLM adapter that wraps any LLM provider to speak the durable protocol.

Reads chat requests from stdin, calls the provided LLM function,
writes responses to stdout.
"""

import json
import sys
from typing import Any, Callable, Dict, List, Optional

from .protocol import ProtocolMessage, read_message, write_message
from .types import Message, ToolCall


class DurableLlmAdapter:
    """Wrap an LLM provider to speak the durable NDJSON protocol.

    Example:
        def my_llm(messages, tools=None, model=None):
            # Call OpenAI, Anthropic, local model, etc.
            return {"content": "Hello!"}

        adapter = DurableLlmAdapter(my_llm)
        adapter.run()
    """

    def __init__(self, handler: Callable):
        """
        handler: A function that takes (messages, tools, model) and returns
                 either {"content": "..."} or {"tool_calls": [...]}.
        """
        self._handler = handler

    def run(self):
        """Run the adapter loop. Reads from stdin, writes to stdout."""
        while True:
            msg = read_message()
            if msg is None:
                break

            if msg.type == "chat":
                messages = msg.payload.get("messages", [])
                tools = msg.payload.get("tools")
                model = msg.payload.get("model")
                self._handle_chat(messages, tools, model)
            elif msg.type == "heartbeat":
                ts = msg.payload.get("timestamp", 0)
                write_message(ProtocolMessage("heartbeat_ack", timestamp=ts))
            else:
                write_message(ProtocolMessage(
                    "error",
                    message=f"unknown message type: {msg.type}",
                    retryable=False,
                ))

    def _handle_chat(self, messages: list, tools: Any, model: Optional[str]):
        try:
            result = self._handler(messages, tools=tools, model=model)

            if isinstance(result, dict):
                if "tool_calls" in result:
                    write_message(ProtocolMessage(
                        "tool_calls",
                        calls=result["tool_calls"],
                    ))
                else:
                    content = result.get("content", result.get("text", ""))
                    write_message(ProtocolMessage("text", content=content))
            elif isinstance(result, str):
                write_message(ProtocolMessage("text", content=result))
            else:
                write_message(ProtocolMessage("text", content=str(result)))

        except Exception as e:
            write_message(ProtocolMessage(
                "error",
                message=str(e),
                retryable=True,
            ))
